### PR TITLE
Move plugin tutorial to Coq repo

### PR DESCRIPTION
**Pull request by SkySkimmer**
There's been discussion about the plugin tutorial fitting better in the main repo, so here's a PR moving it.
I'm not sure if everyone was in agreement on this so putting `needs: discussion`.

I put the code in doc/plugin_tutorial and build it in the test suite. I used git-subtree so commit history is preserved (as you can see from the github view).

Should I make a CODEOWNERS entry for @ybertot?

### CODE REVIEW

**Comment by ejgallego** [comment]
I am not sure this should be built as part of the test suite, I don't like it.

The tutorial is not a test but a part of the doc, or even a regular set of plugins. I would even prefer it is a separate target / it's own opam package [composed].

**Comment by SkySkimmer** [comment]
I don't know how to do such things.
Also we don't really care about building it (it's not intended to be 
used), we just want to know that it compiles so doing it in the test 
suite makes sense IMO.

Gaëtan Gilbert

**Comment by Zimmi48** [comment]
I was expecting this would be part of `plugins/` but well, no strong opinion on the location.

**Comment by Zimmi48** [comment]
It should probably get a mention somewhere in the manual BTW.

**Comment by SkySkimmer** [comment]
I think changing the license can be left to a followup PR.

**Review by ejgallego — COMMENTED** [review]
I'd rather include the plugin tutorial in the main build instead of as a test. The workflow seems way more natural, in particular with relation to dev tools.

Also, the tutorial should be Dunerized. That's very easy but I can do it if you want.

**Comment by SkySkimmer** [comment]
>I'd rather include the plugin tutorial in the main build instead of as a test. The workflow seems way more natural, in particular with relation to dev tools.

I don't really see how, can you expand a bit more?

>Also, the tutorial should be Dunerized. That's very easy but I can do it if you want.

What do you mean by dunerized?

**Comment by SkySkimmer** [comment]
>Add a plugin-tutorial target to the makefile, then have make world build it.

OK, although I meant about how the workflow is more natural.

**Comment by ejgallego** [comment]
Ah ok! Somehow I think of the tutorial as a set of plugins, so it makes sense to me for it to be built together with the rest of the plug-ins. I would indeed expect that we put test of the plugins in the test suite, not the build itself (which yet introduces another ad-hoc testing mechanism)

I don't have a strong opinion on this tho, so if people feel like this should be a test ok.

**Comment by SkySkimmer** [comment]
So are we going to do this or not?

File: doc/plugin_tutorial/tuto2/src/demo.ml4
Comment (jfehrle) [inline doc/plugin_tutorial/tuto2/src/demo.ml4]:
```suggestion
   We have decided to name this plugin "demo_plugin".
```

File: doc/plugin_tutorial/tuto2/src/demo.ml4
Comment (jfehrle) [inline doc/plugin_tutorial/tuto2/src/demo.ml4]:
```suggestion
   As in case of the parsing rule,
```

File: doc/plugin_tutorial/tuto2/src/demo.ml4
Comment (jfehrle) [inline doc/plugin_tutorial/tuto2/src/demo.ml4]:
```suggestion
   That is, it behaves as a no-op.
```

File: doc/plugin_tutorial/tuto2/src/demo.ml4
Comment (jfehrle) [inline doc/plugin_tutorial/tuto2/src/demo.ml4]:
```suggestion
   in those places where we expect users to provide input
```

File: doc/plugin_tutorial/tuto3/src/tuto_tactic.ml
Comment (jfehrle) [inline doc/plugin_tutorial/tuto3/src/tuto_tactic.ml]: like what?

**Review by jfehrle — APPROVED** [review]
Some minor wording tweaks

**Comment by jfehrle** [comment]
Why is the PR named "*Vendor* plugin tutorial" rather than just "Plugin tutorial"?

**Comment by ejgallego** [comment]
> Why is the PR named "_Vendor_ plugin tutorial" rather than just "Plugin tutorial"?

Because the discussion is about vendoring the plugin tutorial.

**Comment by ejgallego** [comment]
> So are we going to do this or not?

I guess @ybertot as the upstream author should say what he thinks.

**Comment by jfehrle** [comment]
> Because the discussion is about vendoring the plugin tutorial.

"vendor" is not a standard English verb.  I don't know what is meant here.

**Comment by SkySkimmer** [comment]
@jfehrle I'm going to mark your diff comments "off topic" as I want this PR to be a clean import from the other repo as much as possible. Improvements on the tutorials should be submitted either at ybertot/plugin_tutorials until we move to merge this, or wait until after we merge this.

**Comment by SkySkimmer** [comment]
I guess it's not vendoring in the strict sense as the intent is to make this the main repo for the tutorials.

**Comment by jfehrle** [comment]
> I'm going to mark your diff comments "off topic" as I want this PR to be a clean import from the other repo as much as possible.

Making them a separate commit in the same PR is not sufficient?  Do what you think is best.

**Comment by gares** [comment]
Discussion happened during the last WG. The decision was to include it, to relicense it under a permissive license called [unlicense](https://choosealicense.com/licenses/unlicense/) (all authors agreed) by email or were recorded on video stream. Unclear were to put it, possibly in `doc/`.

**Review by gares — APPROVED** [review]

**Comment by SkySkimmer** [comment]
>Unclear were to put it, possibly in doc/.

In doc/ is the current state of the PR.

This is my plan:
- merge https://github.com/ybertot/plugin_tutorials/pull/22 (license change)
- lock the plugin_tuto repo (github archive feature or just informally)
- rebase the PR
- merge the PR
- move open issues on the plugin_tuto to coq (actually there aren't except for the licensing issue any so this is a noop)
- tell PR authors to move their PRs
- done

**Comment by Zimmi48** [comment]
I suggest changing CODEOWNERS so that @ybertot is still the main code owner of the plugin tutorial...

**Comment by SkySkimmer** [comment]
>I suggest changing CODEOWNERS so that @ybertot is still the main code owner of the plugin tutorial...

Done

**Comment by Zimmi48** [comment]
```
Makefile:1: warning: undefined variable 'COQBIN'
/coq_makefile -f _CoqProject -o Makefile.coq
make[3]: /coq_makefile: Command not found
```

File: doc/plugin_tutorial/tuto3/theories/Data.v
Comment (SkySkimmer) [inline doc/plugin_tutorial/tuto3/theories/Data.v]: Currently this triggers the template poly warning, let's just wait until #9304 is merged and it will go away.

File: doc/plugin_tutorial/tuto3/theories/Data.v
Comment (SkySkimmer) [inline doc/plugin_tutorial/tuto3/theories/Data.v]: Removed so that `states` is enough to compile the tuto.

**Comment by SkySkimmer** [comment]
Getting coq_makefile to work with a Coq that hasn't been installed yet is a pain.
This is why I wanted to run in the test suite, it's already set up for running coq_makefile.

**Comment by Zimmi48** [comment]
I would be OK to putting this in test-suite as a first step if this is easier to do, with the understanding that it should be moved back to `plugins/` once Dune support is ready.

**Comment by SkySkimmer** [comment]
It makes no sense in plugins/
Anyway I got it to work by using a separate job with configure -local

**Comment by SkySkimmer** [comment]
Should be ready to merge.

**Comment by Zimmi48** [comment]
@ybertot Would you like to be the one to merge this, as the (only) code owner, or should it be someone else?

**Review by maximedenes — APPROVED** [review]

**Comment by ybertot** [comment]
I am merging this, but past history has shown that I am not responsive enough for the speed of development of the team, so there should be another code owner.

**Comment by SkySkimmer** [comment]
@ybertot I think you should lock the plugin_tutorials repo.

**Comment by Zimmi48** [comment]
>@ybertot I think you should lock the plugin_tutorials repo.

Recommended course of action:
- Change the README and the description to indicate the move and the new location.
- Use the "archive" feature of GitHub in Settings which will make the repo read-only.

**Pull request merged**
